File: num_meth/sem5/lab_3/gauss.f
      subroutine gauss (n,matr,extension,solve)
      real, dimension(n,n) :: matr
      real, dimension(n,n+1) :: ext_matr
      real, dimension(n) :: ni,extension,solve

      do j=1,n
         ext_matr(:,j)=matr(:,j)
      end do
      ext_matr(:,n+1)=extension
      do j=1,n
         ni(j)=0
      end do
      nj=n+1
      do ij=1,n
         b=0.
         do i=1,n
            if (ni(i)==0) then
               do j=1,n
                  c=abs(ext_matr(i,j))
                  d=abs(b)
                  if (c>d) then
                     b=ext_matr(i,j)
                     k=i
                     l=j
                  end if
               enddo
            end if
         end do
         ni(k)=l
         ext_matr(k,l)=0.
         d=abs(b)
         if (d==0.0) then
            print *,'Решений Нет!'
            stop
         end if
         do i=1,nj
            c=abs(ext_matr(k,i))
            if (c/=0.0) then
               s=ext_matr(k,i)/b
               ext_matr(k,i)=s
               do j=1,n
                  ext_matr(j,i)=ext_matr(j,i)-ext_matr(j,l)*s
               end do
            end if
         end do
         do j=1,n
            ext_matr(j,l)=0.
         end do
      end do
      do j=1,n
         k=ni(j)
         solve(k)=ext_matr(j,n+1)
      end do
      return
      end subroutine gauss 
      
c$$$      function resid(n, matr, solve, extesion)
c$$$      real, dimension(n,n) :: matr
c$$$      real, dimension(n) :: extesion, solve
c$$$      real, dimension(n) :: resid
c$$$      resid = matmul(matr, solve) - extesion
c$$$      resid = maxval(abs(matmul(matr,solve) - extesion))
c$$$      end function resid
      
      program main
      IMPLICIT NONE
      integer :: n
      parameter (n=3)
      real, dimension(n,n) :: matrix
      real, dimension(n) :: extension, solve, resid
      real :: norm 
!matrix(1,:)=(/0., 0., 0./)
      matrix(1,:)=(/1.80, 2.50, 4.60/)
      matrix(2,:)=(/3.10, 2.30, -1.20/)
      matrix(3,:)=(/4.51, -1.80, 3.60/)
      extension =(/2.20,3.60,-1.70/)    

      call gauss(n, matrix, extension, solve)

      resid = matmul(matrix, solve) - extension
      norm = maxval(abs(resid))

      print*,'X1 = ',solve(1)
      print*,'X2 = ',solve(2)
      print*,'X3 = ',solve(3)
      print*,'resid = (', resid, ')'
      print*,'resid norm = ', norm
      end program main

File: num_meth/sem5/lab_3/gauss.py
def gauss(A, b):
    def _prepare_main_elem():
        for i in range(len(b)):
            for j in range(len(b)):
                if(i != j and A[i][i] == 0 and A[j][i] != 0 and A[i][j] != 0):                    
                    for k in range(len(b)):
                        A[j][k], A[i][k] = A[i][k], A[j][k]
                    b[j],b[i] = b[i],b[j]
                    break

    _prepare_main_elem()
    for i in range(len(b)):
        if A[i][i] == 0:
            return 'e'
        for j in range(i+1, len(b)):
            tmp = A[j][i] / A[i][i]
            for t in range(i, len(b)):
                A[j][t] -= tmp * A[i][t]
            b[j] -= tmp * b[i]
    solve = [0 for i in range(len(b))]
    for i in reversed(range(0, len(b))):
        sum_known_elems = 0
        for j in range(i, len(b)):
            sum_known_elems += A[i][j] * solve[j]
        solve[i] = (b[i] - sum_known_elems) / A[i][i]
    return solve;

def resid(A, b, solve):
    F = []
    for i in range (len(A)):
        res = 0
        for j in range (len(solve)):
            res += A[i][j] * solve[j]
        F.append(res - b[i])
    return F

M = [[1.80, 2.50, 4.60],
     [3.10, 2.30, -1.20],
     [4.51,-1.80,3.60]]
c = [2.20,3.60,-1.70]

solve = gauss(M,c)
if solve == 'e':
    print("Решений Нет!")
else:
    [print('X{0} = {1}'.format(i, solve[i])) for i in range(len(c))]
    residial = resid(M, c, solve)
    print('resid = ({0})'.format(residial))
    norm = max([abs(residial[i]) for i in range(len(residial))])
    print('resid norm = {0}'.format(norm))


